newbank/server/LoanManager.java

```diff
@@ -0,0 +1,186 @@
+package newbank.server;
+
+import java.util.HashMap;
+
+public class LoanManager {
+  private static final LoanManager loanManager = new LoanManager();
+  private static final NewBankQuery_I bank = NewBank.getBank();
+
+  private HashMap<Integer, Loan> loans;
+  private int lastLoanNumber;
+
+  private LoanManager() {
+    loans = new HashMap<>();
+    lastLoanNumber = 0;
+  }
+
+  public static LoanManager getLoanManager() {
+    return loanManager;
+  }
+
+  public void clearLoans() {
+    this.loans.clear();
+    lastLoanNumber = 0;
+  }
+  
+  public String showTakenLoans(CustomerID customerID) {
+    Customer customer = bank.getCustomer(customerID);
+    String result = "";
+
+    for (Account account : customer.getAllAccounts()) {
+      for (int loanID : loans.keySet()) {
+        Loan loan = loans.get(loanID);
+        if (loan.getAccountTo() != null
+            && loan.getAccountTo().getAccountNumber().equals(account.getAccountNumber())) {
+          result = result + "Loan Number: " + loan.getNumber() + ", Account Number: "
+              + loan.getAccountFrom().getAccountNumber() + ", Amount: " + loan.getAmount()
+              + ", Interest Rate: " + loan.getInterest() + "%, Taken by: "
+              + loan.getAccountTo().getAccountNumber() + "\n";
+        }
+      }
+    }
+    return result == "" ? "No loans taken" : result;
+  }
+
+  public String paybackLoan(CustomerID customerID, int loanNumber) {
+    Loan loan = loans.get(loanNumber);
+    if (loan == null) {
+      return "Error. Invalid loan number.";
+    } else {
+      Account accountTo = loan.getAccountFrom();
+      Account accountFrom = loan.getAccountTo();
+      if (accountTo == null) {
+        return "Error. Invalid account number.";
+      }
+      if (accountFrom == null) {
+        return "Error. Invalid account number.";
+      }
+
+      return handleLoanPayment(loan, loan.getAmount());
+    }
+  }
+
+  public String paybackLoanPartial(CustomerID customerID, int loanNumber, double value) {
+    Loan loan = loans.get(loanNumber);
+    if (loan == null) {
+      return "Error. Invalid loan number.";
+    } else {
+      Account accountTo = loan.getAccountFrom();
+      Account accountFrom = loan.getAccountTo();
+      if (accountTo == null) {
+        return "Error. Invalid account number.";
+      }
+      if (accountFrom == null) {
+        return "Error. Invalid account number.";
+      }
+
+      return handleLoanPayment(loan, value);
+    }
+  }
+
+  private String handleLoanPayment(Loan loan, double value) {
+    Account accountTo = loan.getAccountFrom();
+    Account accountFrom = loan.getAccountTo();
+    double paymentAmount = value;
+    double loanAmount = loan.getAmount();
+
+    if (value <= loanAmount) {
+      loan.pay(value);
+    } else {
+      // loans.remove(loan.getNumber());
+      paymentAmount = loanAmount;
```

Comment:
Youre right, initially we wanted to remove the loan when it was completed, so we need to decide whether to do this or pay the loan amount :)

Reply:
Maybe a good idea would be to keep the loan but give it a status of 'Closed' or something? That way the system keeps a record of previous loans rather than them being removed

Reply:
The problem with this is that the program will grow in size until it runs out of memory

Reply:
Oooh yeah that's true! Maybe we add a task to the backlog for the system to delete any loans older than 2 years or something?

Reply:
I have branched this branch and added an 'openLoan' boolean which starts at true and is toggled to false when the loan is paid off.

Reply:
Sounds good to me!